Make JSExtractor serialize iteratively, restoring the ability to serialize object graphs 40,000 levels deep.

beidson opened the pull request:
#### 4b8d7f51936cbbc9b16d36278a0ac2c9b50439e1
<pre>
Make JSExtractor serialize iteratively, restoring the ability to serialize object graphs 40,000 levels deep.
<a href="https://rdar.apple.com/164658717">rdar://164658717</a>

Reviewed by Geoffrey Garen.

When serialized JavaScript objects to send to the UI process, we used to use SerializedScriptValue.
It walked JS object graphs iteratively, with a preset limit of going 40,000 nesting levels deep.

When we moved entirely off of SerializedScriptValue and onto JSExtractor we changed from iterative to
recursive behavior, therefore limiting our nesting depth to whatever the stack could handle.

Originally this meant particularly deep object graphs could exhaust the stack and crash the WebContent
process, which we detected and stopped in <a href="https://commits.webkit.org/303006@main">https://commits.webkit.org/303006@main</a>

But this still left us in a situation where a JS object that might have successfully serialized before
would now fail to serialize. Informally I found about ~12,500 nesting levels deep instead of 40,000.

We know of at least one WKWebView app that evaluates JavaScript and was falling victim to this.

This patch restores iterative behavior to JSExtractor, removing the reliance on the system stack.
It also adds the same 40,000 deep restriction that SerializedScriptValue.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewEvaluateJavaScript.mm

* Source/WebKit/Shared/JavaScriptEvaluationResult.cpp:
(WebKit::JavaScriptEvaluationResult::JSExtractor::addObjectToMap): Instead of being part of a recursive
  algorithm, this function now iteratively walks the entire object graph, one nesting level at a time.
(WebKit::JavaScriptEvaluationResult::JSExtractor::extractJSValue):
(WebKit::JavaScriptEvaluationResult::JSExtractor::jsValueToExtractedValue): Directly extracts the current
  JSValueRef into a serialized Value. In the case of Arrays or Objects that need identifiers from values
  on the next nesting level, it keeps a placeholder of their structure to be filled in at the end.
(WebKit::JavaScriptEvaluationResult::JSExtractor::isSafeToRecurse): Deleted.
(WebKit::JavaScriptEvaluationResult::JSExtractor::toValue): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewEvaluateJavaScript.mm:
((EvaluateJavaScript, Serialization)):

Canonical link: <a href="https://commits.webkit.org/303074@main">https://commits.webkit.org/303074@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b9cb744de59c21cdb36534b1734cfd321a340d75

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131159 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/3488 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42123 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138601 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/82878 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/bf085751-4f84-49bc-b3be-59372e580341) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/133029 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/3481 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/3330 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99930 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/67693 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7439f550-3798-468e-bedf-b17491061c56) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134105 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/2488 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/117412 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80629 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d525db4a-a65c-47a6-98fb-34691e5ef0d3) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2408 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/96 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81848 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110995 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/35530 "Found 1 new test failure: inspector/page/hidpi-snapshot-size.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/141097 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/3233 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36055 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/108448 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/3279 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2890 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/108390 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27560 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2420 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113744 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/56288 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/3296 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32168 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/3118 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/66703 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/3318 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/3226 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebKit/Shared/JavaScriptEvaluationResult.cpp:
```diff
@@ -308,8 +355,37 @@ std::optional<JavaScriptEvaluationResult> JavaScriptEvaluationResult::extract(JS
+    }, [&] (UnsupportedTypeTag) {
+        // We already committed an entry into the values map for this object identifier,
+        // but it failed to extract. So remove it from the values map.
+        m_valuesInMap.remove(addResult.iterator);
```
We just ran arbitrary code, so we should not reuse this iterator.

geoffreygaren commented on Source/WebKit/Shared/JavaScriptEvaluationResult.cpp:
```diff
@@ -252,41 +267,73 @@ RefPtr<API::Object> JavaScriptEvaluationResult::toAPI()
-    return !m_didStackOverflow;
-}
+        RELEASE_ASSERT(m_currentJSValuesToExtract.isEmpty());
+        m_currentJSValuesToExtract = WTFMove(m_futureJSValuesToExtract);
```
Though Chris has worked to make use after move safe, it's probably still not best practice. Let's do `std::exchange(m_futureJSValuesToExtract, { })`.

geoffreygaren commented on Source/WebKit/Shared/JavaScriptEvaluationResult.cpp:
```diff
@@ -308,8 +355,37 @@ std::optional<JavaScriptEvaluationResult> JavaScriptEvaluationResult::extract(JS
+    if (m_valuesInMap.contains({ context, value }))
+        return;
+    if (m_currentJSValuesToExtract.contains({ context, value }))
+        return;
```
I don't love it that we have to be careful here. Would it work to unconditionally add to the future values map, and check later, when we encounter the value, whether it's already in m_valuesInMap?

That way, the check that happens later can just be an early return instead of a crash. And there's less code. And possibly fewer hash lookups.

geoffreygaren commented on Source/WebKit/Shared/JavaScriptEvaluationResult.cpp:
```diff
@@ -308,8 +355,37 @@ std::optional<JavaScriptEvaluationResult> JavaScriptEvaluationResult::extract(JS
+{
+    auto identifier = JSObjectID::generate();
+    auto addResult = m_valuesInMap.add({ context, value }, identifier);
+    RELEASE_ASSERT(addResult.isNewEntry);
```
See other comment about returning early instead of crashing.

geoffreygaren replied:
Specifically, you can call `ensure()` here, so that you only generate the ID if you are going to add to the map; else return early.

geoffreygaren commented on Source/WebKit/Shared/JavaScriptEvaluationResult.cpp:
```diff
@@ -55,15 +55,30 @@ class JavaScriptEvaluationResult::JSExtractor {
+    size_t m_currentNestingLevel { 0 };
+
+    HashSet<Protected<JSValueRef>> m_futureJSValuesToExtract;
+    HashSet<Protected<JSValueRef>> m_currentJSValuesToExtract;
```
I'm pretty sure you can combine these into a single `Vector<Protected<JSValueRef>>`. I believe they are only hash tables for the sake of the uniqueness check, and if you agree that the uniqueness check can be removed (and instead done later, when we pop a value off the stack I am proposing), then this can become simpler and more efficient.

beidson replied:
We chatted on Slack about how these two different collections make it clear on maintaining the nesting level.
I'll be changing them to Vectors, though!

geoffreygaren reviewed (changes requested):
Look pretty good but I think I spotted a bug or two and I have a design suggestion.

beidson committed:
Make JSExtractor serialize iteratively, restoring the ability to serialize object graphs 40,000 levels deep.
rdar://164658717

Reviewed by Geoffrey Garen.

When serialized JavaScript objects to send to the UI process, we used to use SerializedScriptValue.
It walked JS object graphs iteratively, with a preset limit of going 40,000 nesting levels deep.

When we moved entirely off of SerializedScriptValue and onto JSExtractor we changed from iterative to
recursive behavior, therefore limiting our nesting depth to whatever the stack could handle.

Originally this meant particularly deep object graphs could exhaust the stack and crash the WebContent
process, which we detected and stopped in https://commits.webkit.org/303006@main

But this still left us in a situation where a JS object that might have successfully serialized before
would now fail to serialize. Informally I found about ~12,500 nesting levels deep instead of 40,000.

We know of at least one WKWebView app that evaluates JavaScript and was falling victim to this.

This patch restores iterative behavior to JSExtractor, removing the reliance on the system stack.
It also adds the same 40,000 deep restriction that SerializedScriptValue.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewEvaluateJavaScript.mm

* Source/WebKit/Shared/JavaScriptEvaluationResult.cpp:
(WebKit::JavaScriptEvaluationResult::JSExtractor::addObjectToMap): Instead of being part of a recursive
  algorithm, this function now iteratively walks the entire object graph, one nesting level at a time.
(WebKit::JavaScriptEvaluationResult::JSExtractor::extractJSValue):
(WebKit::JavaScriptEvaluationResult::JSExtractor::jsValueToExtractedValue): Directly extracts the current
  JSValueRef into a serialized Value. In the case of Arrays or Objects that need identifiers from values
  on the next nesting level, it keeps a placeholder of their structure to be filled in at the end.
(WebKit::JavaScriptEvaluationResult::JSExtractor::isSafeToRecurse): Deleted.
(WebKit::JavaScriptEvaluationResult::JSExtractor::toValue): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewEvaluateJavaScript.mm:
((EvaluateJavaScript, Serialization)):

Canonical link: https://commits.webkit.org/303074@main
